MBS-9085 [Configuration Cache] support buildChecks

dsvoronin opened the pull request:
1. RequiredPluginsChecker removed
heavy subprojects call and not viable after convention plugins

2. IncrementalKaptChecks removed
heavy subprojects call, implemented as part of room convention

3. DynamicDependenciesCheck removed
heavy subprojects call, covered with version lock

dsvoronin committed:
MBS-9085 buildChecks: support configuration cache

1. RequiredPluginsChecker removed
heavy subprojects call and not viable after convention plugins

2. IncrementalKaptChecks removed
heavy subprojects call, implemented as part of room convention

3. DynamicDependenciesCheck removed
heavy subprojects call, covered with version lock

eugene-krivobokov commented on subprojects/gradle/build-checks/src/gradleTest/kotlin/com/avito/android/build_checks/ConfigurationCacheCompatibilityTest.kt:
```diff
@@ -33,7 +39,9 @@ internal class ConfigurationCacheCompatibilityTest {
             projectDir,
             "help",
-            dryRun = true,
+            "-Pkapt.incremental.apt=true",
```
Why do we need this? How it affects the build check?

dsvoronin replied:
no more, when i extracted room check

eugene-krivobokov replied:
Sorry, still didn't get why we use this flag here.

dsvoronin replied:
it was used with different set of tasks to check "configuration cache", i just forgot to remove it

eugene-krivobokov commented on subprojects/gradle/build-checks/src/main/kotlin/com/avito/android/build_checks/BuildParamCheckPlugin.kt:
```diff
@@ -58,7 +57,8 @@ public open class BuildParamCheckPlugin : Plugin<Project> {
 
-        val logger = GradleLoggerFactory.getLogger(this, project)
+        val loggerFactory = GradleLoggerFactory.fromProject(project)
+        val logger = loggerFactory.create(this::class.java.simpleName)
```
Now clients have to do these two operations instead of one request.
How about hiding it inside factory?

dsvoronin replied:
done in https://github.com/avito-tech/avito-android/pull/1198

eugene-krivobokov commented on subprojects/gradle/build-checks/src/main/kotlin/com/avito/android/build_checks/internal/incremental_kapt/RoomIncrementalKaptChecker.kt:
```diff
@@ -1,80 +0,0 @@
-import java.net.URLClassLoader
-import javax.annotation.processing.AbstractProcessor
-
-internal class RoomIncrementalKaptChecker(
```
> IncrementalKaptChecks removed
> heavy subprojects call, implemented as part of room convention

I didn't find equivalent checking logic in room convention plugin.

eugene-krivobokov replied:
I see, it's in another PR. 

I propose to have it in this plugin because of:
- Integration tests
- OS useful plugin

dsvoronin replied:
it should be separate plugin, because it is intended to be applied to modules, not a root project

eugene-krivobokov replied:
Discussed that we can remove this check due to Java 11 wider adoption.

eugene-krivobokov commented:
Please, update the documentation accordingly

eugene-krivobokov commented on subprojects/gradle/build-checks/src/main/kotlin/com/avito/android/build_checks/RootProjectChecksExtension.kt:
```diff
@@ -28,6 +28,7 @@ public open class RootProjectChecksExtension : BuildChecksExtension() {
     public fun gradleDaemon(action: Action<GradleDaemon>): Unit =
         register(GradleDaemon(), action)
 
+    @Deprecated("remove it, has no effect, use versions lock")
```
or ResolutionStrategy

eugene-krivobokov commented on subprojects/gradle/build-checks/src/main/kotlin/com/avito/android/build_checks/internal/CheckGradleDaemonTask.kt:
```diff
@@ -3,15 +3,19 @@ package com.avito.android.build_checks.internal
 import java.io.File
 import java.util.Properties
 
 internal abstract class CheckGradleDaemonTask : DefaultTask() {
```
We could remove this check. The implementation is useless with convention plugins.
Need to redesign it completely.

dsvoronin replied:
will remove

dsvoronin committed:
MBS-9085 [Configuration Cache] remove logger factory methods access to project  (#1198)

* MBS-9085 buildChecks: support configuration cache

1. RequiredPluginsChecker removed
heavy subprojects call and not viable after convention plugins

2. IncrementalKaptChecks removed
heavy subprojects call, implemented as part of room convention

3. DynamicDependenciesCheck removed
heavy subprojects call, covered with version lock

* MBS-9085 remove GradleLoggerFactory task factory methods

because it was called on task execution time and accessing project which is blocks configuration cache

* MBS-9085 [Configuration cache] sentry support (#1199)

* MBS-9085 sentry buildFinished listener removed

configuration cache support

* MBS-9085 buildVerdict was crashing in disabled state

* MBS-9085 remove useless UnstableApiUsage

dsvoronin committed:
MBS-9085 update docs, remove DaemonCheck

eugene-krivobokov commented on subprojects/gradle/build-checks/src/main/kotlin/com/avito/android/build_checks/RootProjectChecksExtension.kt:
```diff
@@ -38,11 +39,11 @@ public open class RootProjectChecksExtension : BuildChecksExtension() {
 
-    @Deprecated("remove it, has no effect, use versions lock")
+    @Deprecated("remove it, has no effect")
     public fun incrementalKapt(action: Action<IncrementalKapt>): Unit =
```
We keep extensions to preserve backward compatibility for clients.
Deprecated methods are easy to miss, especially in groovy files.
It'd be good to log also a warning about usage of deprecated API like AGP and other plugins do.

dsvoronin replied:
done

eugene-krivobokov commented on build-logic/publication/src/main/kotlin/com/avito/android/publish/AndroidLibraryPublishExtension.kt:
```diff
@@ -12,7 +12,6 @@ abstract class AndroidLibraryPublishExtension @Inject constructor(objects: Objec
      *
      * default: release
      */
-    @Suppress("UnstableApiUsage")
```
I'd love to see such massive changes in a separate PR.
Separate commits don't solve all inconveniences.

dsvoronin replied:
Ok, i'll keep that in mind;
i thought it was just a boyscout refactoring, like remove some meta info from files

dsvoronin replied:
done

dsvoronin committed:
MBS-9085 fix fromTask/createstaingrepo

dsvoronin committed:
MBS-9085 fix detekt
